src/file_io.rs

```diff
@@ -0,0 +1,119 @@
+use std::{fs, io, iter::Iterator};
+
+pub trait FsReadWriter {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir>;
+}
+
+pub struct Fs {}
+
+impl FsReadWriter for Fs {
+    fn read_dir(&self, path: &str) -> io::Result<fs::ReadDir> {
+        fs::read_dir(path)
+    }
+}
+
+pub struct FilterResult {
+    pub paths: Vec<String>,
+    pub io_errors: Vec<io::Error>,
+}
+
+pub struct FsFilterer<FS: FsReadWriter> {
+    fs_rw: FS,
+}
+
+impl FsFilterer<Fs> {
+    pub fn new() -> FsFilterer<Fs> {
+        FsFilterer { fs_rw: Fs {} }
+    }
+}
+
+impl<FS: FsReadWriter> FsFilterer<FS> {
+    pub fn new_custom(fs_rw: FS) -> FsFilterer<FS> {
```

Comment:
I'm assuming you can't overload w Rust?

Reply:
No, and I really hate how i ended up doing this, just don't know any other simple way to inject the dependency.